### Machine Learning Algorithm

### Import the relevant libraries

In [2]:
import numpy as np
import tensorflow as tf

### Data

In [3]:
# let's create a temporary variable npz, where we will store each of the three Audiobooks datasets
npz = np.load('Audiobooks_data_train.npz')

# we extract the inputs using the keyword under which we saved them
# to ensure that they are all floats, let's also take care of that
train_inputs = npz['inputs'].astype(np.float64)
# targets must be int because of sparse_categorical_crossentropy (we want to be able to smoothly one-hot encode them)
train_targets = npz['targets'].astype(np.int64)

# we load the validation data in the temporary variable
npz = np.load('Audiobooks_data_validation.npz')
# we can load the inputs and the targets in the same line
validation_inputs, validation_targets = npz['inputs'].astype(np.float64), npz['targets'].astype(np.int64)

# we load the test data in the temporary variable
npz = np.load('Audiobooks_data_test.npz')
# we create 2 variables that will contain the test inputs and the test targets
test_inputs, test_targets = npz['inputs'].astype(np.float64), npz['targets'].astype(np.int64)

### Model
Outline, optimizers, loss, early stopping and training

In [43]:
# Set the input and output sizes
input_size = 10
output_size = 2
# Use same hidden layer size for both hidden layers. Not a necessity.
hidden_layer_size = 50
    
# define how the model will look like
model = tf.keras.Sequential([
    # tf.keras.layers.Dense is basically implementing: output = activation(dot(input, weight) + bias)
    # it takes several arguments, but the most important ones for us are the hidden_layer_size and the activation function
    tf.keras.layers.Dense(hidden_layer_size, activation='relu'), # 1st hidden layer
    tf.keras.layers.Dense(hidden_layer_size, activation='relu'), # 2nd hidden layer
    # the final layer is no different, we just make sure to activate it with softmax
    tf.keras.layers.Dense(output_size, activation='softmax') # output layer
])


### Choose the optimizer and the loss function

# we define the optimizer we'd like to use, 
# the loss function, 
# and the metrics we are interested in obtaining at each iteration
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])

### Training
# That's where we train the model we have built.

# set the batch size
batch_size = 100

# set a maximum number of training epochs
max_epochs = 100


# fit the model
# note that this time the train, validation and test data are not iterable
model.fit(train_inputs, # train inputs
          train_targets, # train targets
          batch_size=batch_size, # batch size
          epochs=max_epochs, # epochs that we will train for (assuming early stopping doesn't kick in)
          validation_data=(validation_inputs, validation_targets), # validation data
          verbose = 2 # making sure we get enough information about the training process
          )  

Epoch 1/100
36/36 - 3s - 93ms/step - accuracy: 0.6689 - loss: 0.5960 - val_accuracy: 0.7897 - val_loss: 0.4663
Epoch 2/100
36/36 - 0s - 11ms/step - accuracy: 0.7471 - loss: 0.4670 - val_accuracy: 0.8210 - val_loss: 0.3940
Epoch 3/100
36/36 - 0s - 12ms/step - accuracy: 0.7762 - loss: 0.4210 - val_accuracy: 0.8389 - val_loss: 0.3559
Epoch 4/100
36/36 - 0s - 11ms/step - accuracy: 0.7815 - loss: 0.3932 - val_accuracy: 0.7740 - val_loss: 0.3576
Epoch 5/100
36/36 - 0s - 11ms/step - accuracy: 0.7958 - loss: 0.3802 - val_accuracy: 0.8523 - val_loss: 0.3379
Epoch 6/100
36/36 - 0s - 11ms/step - accuracy: 0.8033 - loss: 0.3687 - val_accuracy: 0.8635 - val_loss: 0.3154
Epoch 7/100
36/36 - 0s - 11ms/step - accuracy: 0.8041 - loss: 0.3614 - val_accuracy: 0.8389 - val_loss: 0.3224
Epoch 8/100
36/36 - 0s - 11ms/step - accuracy: 0.8044 - loss: 0.3580 - val_accuracy: 0.8345 - val_loss: 0.3174
Epoch 9/100
36/36 - 0s - 11ms/step - accuracy: 0.8030 - loss: 0.3565 - val_accuracy: 0.8568 - val_loss: 0.3125
E

## Test the model

We test the final prediction power of our model by running it on the test dataset that the algorithm has NEVER seen before.

It is very important to realize that fiddling with the hyperparameters overfits the validation dataset. 

The test is the absolute final instance. You should not test before you are completely done with adjusting your model.

If you adjust your model after testing, you will start overfitting the test dataset, which will defeat its purpose.

In [49]:
test_loss, test_accuracy = model.evaluate(test_inputs, test_targets)

14/14 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.8596 - loss: 0.2895 


In [51]:
print('\nTest accuracy: {0:.2f}%'.format( test_accuracy*100.))


Test accuracy: 84.38%


Using the initial model and hyperparameters given in this notebook, the final test accuracy should be roughly around 85%.
